modules/programs/streamlink.nix

```diff
@@ -0,0 +1,143 @@
+
+    settings = mkOption {
+      type = with types;
+        attrsOf (oneOf [ bool int str (listOf (either int str)) ]);
+      default = { };
+      example = literalExpression ''
+        {
+          player = "''${pkgs.mpv}/bin/mpv";
+          player-args = "--cache 2048";
+          player-no-close = true;
+        }
+      '';
+      description = ''
+        Global configuration options for streamlink. It will be written to
+        {file}`$XDG_CONFIG_HOME/streamlink/config` (linux) or
+        {file}`Library/Application Support/streamlink/config` (darwin).
+      '';
+    };
+
+    plugins = mkOption {
+      description = ''
+        Streamlink plugins.
+
+        If a source is set, the custom plugin will be linked to the data directory.
+
+        Additional configuration specific to the plugin, if defined, will be
+        written to the config directory, and override global settings.
+      '';
+      type = types.attrsOf pluginType;
+      default = { };
+      example = literalExpression ''
+        {
+          custom_plugin = {
+            src = ./custom_plugin.py;
+            settings = {
+              quiet = true;
+            };
+          };
+        }
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    home.packages = [ cfg.package ];
+
+    home.file = mkIf pkgs.stdenv.hostPlatform.isDarwin {
+      "Library/Application Support/streamlink/config" =
+        mkIf (cfg.settings != { }) {
+          text = concatStringsSep "\n" (remove "" (renderSettings cfg.settings))
+            + "\n";
+        };
+    } // (mapAttrs' (name: value:
+      nameValuePair "Library/Application Support/streamlink/config.${name}"
+      (mkIf (value.settings != { }) {
+        text = concatStringsSep "\n" (remove "" (renderSettings value.settings))
+          + "\n";
+      })) cfg.plugins) // (mapAttrs' (name: value:
+        nameValuePair
+        "Library/Application Support/streamlink/plugins/${name}.py"
+        (mkIf (value.src != null) (if (builtins.isPath value.src) then {
+          source = value.src;
+        } else {
+          text = value.src;
+        }))) cfg.plugins);
+
+    xdg.configFile = mkIf pkgs.stdenv.hostPlatform.isLinux ({
+      "streamlink/config" = mkIf (cfg.settings != { }) {
+        text = concatStringsSep "\n" (remove "" (renderSettings cfg.settings))
+          + "\n";
+      };
+    } // (mapAttrs' (name: value:
+      nameValuePair "streamlink/config.${name}" (mkIf (value.settings != { }) {
+        text = concatStringsSep "\n" (remove "" (renderSettings value.settings))
+          + "\n";
+      })) cfg.plugins));
+
+    xdg.dataFile = mkIf pkgs.stdenv.hostPlatform.isLinux (mapAttrs'
+      (name: value:
+        nameValuePair "streamlink/plugins/${name}.py" (mkIf (value.src != null)
+          (if (builtins.isPath value.src) then {
+            source = value.src;
+          } else {
+            text = value.src;
+          }))) cfg.plugins);
+  };
```

Comment:
Would be easier to read if we use `mkMerge` and a `lib.generators` function. Would also recommend moving the config  path out into a variable outside of the scope of this section so you don't need to have the check inside here on the path.